File: yogi-cpp/test/public_api_tests/timer_test.cc
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2019 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

#include "../common.h"

#include <yogi_core.h>

class TimerTest : public testing::Test {
 protected:
  yogi::ContextPtr context_ = yogi::Context::Create();
  yogi::TimerPtr timer_ = yogi::Timer::Create(context_);
};

TEST_F(TimerTest, Start) {
  bool called = false;
  timer_->Start(yogi::Duration::FromMilliseconds(1), [&](auto& res) {
    EXPECT_NO_THROW(dynamic_cast<const yogi::Success&>(res));
    EXPECT_EQ(res.GetErrorCode(), yogi::ErrorCode::kOk);
    called = true;
  });

  while (!called) {
    context_->RunOne();
  }

  EXPECT_TRUE(called);
}

TEST_F(TimerTest, Cancel) {
  EXPECT_FALSE(timer_->Cancel());

  bool called = false;
  timer_->Start(yogi::Duration::kInfinity, [&](auto& res) {
    EXPECT_NO_THROW(dynamic_cast<const yogi::Failure&>(res));
    EXPECT_EQ(res.GetErrorCode(), yogi::ErrorCode::kCanceled);
    called = true;
  });

  EXPECT_TRUE(timer_->Cancel());

  while (!called) {
    context_->RunOne();
  }

  EXPECT_TRUE(called);
}

File: yogi-dotnet/test/TimerTest.cs
/*
 * This file is part of the Yogi distribution https://github.com/yohummus/yogi.
 * Copyright (c) 2019 Johannes Bergmann.
 *
 * This library is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 3.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this library. If not, see <http://www.gnu.org/licenses/>.
 */

using System;
using System.Net;
using Xunit;

namespace test
{
    public class TimerTest : TestCase
    {
        Yogi.Context context = new Yogi.Context();

        [Fact]
        public void Start()
        {
            Yogi.Timer timer = new Yogi.Timer(context);

            bool called = false;
            timer.StartAsync(Yogi.Duration.FromMilliseconds(1), (res) => {
                Assert.IsType<Yogi.Success>(res);
                Assert.Equal(Yogi.ErrorCode.Ok, res.ErrorCode);
                called = true;
            });

            GC.Collect();

            while (!called)
            {
                context.RunOne();
            }

            Assert.True(called);

            GC.KeepAlive(timer);
        }

        [Fact]
        public void Cancel()
        {
            Yogi.Timer timer = new Yogi.Timer(context);

            Assert.False(timer.Cancel());

            bool called = false;
            timer.StartAsync(Yogi.Duration.Infinity, (res) => {
                Assert.IsType<Yogi.Failure>(res);
                Assert.Equal(Yogi.ErrorCode.Canceled, res.ErrorCode);
                called = true;
            });

            GC.Collect();
            Assert.True(timer.Cancel());
            GC.Collect();

            while (!called)
            {
                context.RunOne();
            }

            Assert.True(called);

            GC.KeepAlive(timer);
        }
    }
}
